dash-pipeline/dpapp/dash/node.c

```diff
@@ -0,0 +1,354 @@
+    DASH_N_ERROR,
+} dash_error_t;
+
+static char *dash_error_strings[] = {
+#define _(sym,string) string,
+  foreach_dash_error
+#undef _
+};
+
+typedef enum
+{
+  DASH_NEXT_INTERFACE_OUTPUT,
+  DASH_N_NEXT,
+} dash_next_t;
+
+#define foreach_mac_address_offset              \
+_(0)                                            \
+_(1)                                            \
+_(2)                                            \
+_(3)                                            \
+_(4)                                            \
+_(5)
+
+VLIB_NODE_FN (dash_node) (vlib_main_t * vm, vlib_node_runtime_t * node,
+			    vlib_frame_t * frame)
+{
+  u32 n_left_from, *from, *to_next;
+  dash_next_t next_index;
+  u32 pkts_swapped = 0;
+
+  from = vlib_frame_vector_args (frame);
+  n_left_from = frame->n_vectors;
+  next_index = node->cached_next_index;
+
+  while (n_left_from > 0)
+    {
+      u32 n_left_to_next;
+
+      vlib_get_next_frame (vm, node, next_index, to_next, n_left_to_next);
+
+      while (n_left_from >= 4 && n_left_to_next >= 2)
+	{
+	  u32 next0 = DASH_NEXT_INTERFACE_OUTPUT;
+	  u32 next1 = DASH_NEXT_INTERFACE_OUTPUT;
+	  u32 sw_if_index0, sw_if_index1;
+	  u8 tmp0[6], tmp1[6];
+	  ethernet_header_t *en0, *en1;
+	  dash_header_t *dh0, *dh1;
+	  u32 bi0, bi1;
+	  vlib_buffer_t *b0, *b1;
+
+	  /* Prefetch next iteration. */
+	  {
+	    vlib_buffer_t *p2, *p3;
+
+	    p2 = vlib_get_buffer (vm, from[2]);
+	    p3 = vlib_get_buffer (vm, from[3]);
+
+	    vlib_prefetch_buffer_header (p2, LOAD);
+	    vlib_prefetch_buffer_header (p3, LOAD);
+
+	    clib_prefetch_store (p2->data);
+	    clib_prefetch_store (p3->data);
+	  }
+
+	  /* speculatively enqueue b0 and b1 to the current next frame */
+	  to_next[0] = bi0 = from[0];
+	  to_next[1] = bi1 = from[1];
+	  from += 2;
+	  to_next += 2;
+	  n_left_from -= 2;
+	  n_left_to_next -= 2;
+
+	  b0 = vlib_get_buffer (vm, bi0);
+	  b1 = vlib_get_buffer (vm, bi1);
+
+	  dh0 = vlib_buffer_get_current (b0);
+	  dh1 = vlib_buffer_get_current (b1);
+	  dash_flow_table_t *flow_table = dash_flow_table_get();
+	  if (dash_flow_process(flow_table, dh0) != 0) {
+	      /* FIXME */
+	  }
+	  if (dash_flow_process(flow_table, dh1) != 0) {
+	      /* FIXME */
+	  }
+
+	  /*
+	   * Build new packet
+	   */
+
+	  /* Update ethernet header via swap src and dst mac */
+	  en0 = (ethernet_header_t*)dh0 - 1;
+#define _(a) tmp0[a] = en0->src_address[a];
+	  foreach_mac_address_offset;
+#undef _
+#define _(a) en0->src_address[a] = en0->dst_address[a];
+	  foreach_mac_address_offset;
+#undef _
+#define _(a) en0->dst_address[a] = tmp0[a];
+	  foreach_mac_address_offset;
+#undef _
+
+	  en1 = (ethernet_header_t*)dh1 - 1;
+#define _(a) tmp1[a] = en1->src_address[a];
+	  foreach_mac_address_offset;
+#undef _
+#define _(a) en1->src_address[a] = en1->dst_address[a];
+	  foreach_mac_address_offset;
+#undef _
+#define _(a) en1->dst_address[a] = tmp1[a];
+	  foreach_mac_address_offset;
+#undef _
+
+	  /* Update dash header */
```

Comment:
will be better to move this into dedicated functions, same as build new packet, verify, build new packet.

Reply:
fixed